lib/models/room.js

```diff
@@ -119,11 +138,71 @@ function Room(roomId, opts) {
         //   data: <receipt data>
         // }]
     };
+
     this._notificationCounts = {};
+
+    // just a list - *not* ordered.
+    this._timelines = [];
+    this._eventIdToTimeline = {};
+    this._timelineSupport = Boolean(opts.timelineSupport);
+
+    this.resetLiveTimeline();
 }
 utils.inherits(Room, EventEmitter);
 
 /**
+ * Get the live timeline for this room.
+ *
+ * @return {module:models/event-timeline~EventTimeline} live timeline
+ */
+Room.prototype.getLiveTimeline = function() {
+    return this._liveTimeline;
+};
+
+/**
+ * Reset the live timeline, and start a new one.
+ *
+ * <p>This is used when /sync returns a 'limited' timeline.
+ */
+Room.prototype.resetLiveTimeline = function() {
+    var newTimeline;
+
+    if (!this._timelineSupport) {
+        // if timeline support is disabled, forget about the old timelines
+        newTimeline = new EventTimeline(this.roomId);
+        this._timelines = [newTimeline];
+        this._eventIdToTimeline = {};
+    } else {
+        newTimeline = this.addTimeline();
+    }
+
+    if (this._liveTimeline) {
+        // initialise the state in the new timeline from our last known state
+        newTimeline.initialiseState(this._liveTimeline.getState(false).events);
+    }
+    this._liveTimeline = newTimeline;
+
+    // maintain this.timeline as a reference to the live timeline,
+    // and this.oldState and this.currentState as references to the
+    // state at the start and end of that timeline. These are more
+    // for backwards-compatibility than anything else.
+    this.timeline = this._liveTimeline.getEvents();
+    this.oldState = this._liveTimeline.getState(true);
+    this.currentState = this._liveTimeline.getState(false);
+};
+
+/**
+ * Get the timeline which contains the given event, if any
+ *
+ * @param {string} eventId  event ID to look for
+ * @return {?module:models/event-timeline~EventTimeline} timeline containing
+ * the given event, or undefined if unknown
```

Comment:
Please return `null` rather than `undefined`. The JS SDK consistently returns `null` to represent "no value exists". `undefined` shouldn't be a part of the public API. By doing this consistently we help devs because if they get an `undefined` error they know it is most likely due to a typo and not due to lack of a null guard.